gprofiler/spark/sampler.py

```diff
@@ -259,30 +66,26 @@ def _collect_loop(self) -> None:
         timefn = time.monotonic
         start_time = timefn()
         while not self._stop_event.is_set():
-            if self._spark_sampler is None:
-                spark_cluster_conf = self._find_spark_cluster()
-                if spark_cluster_conf is not None:
-                    master_address, cluster_mode = spark_cluster_conf
-                    self._spark_sampler = SparkCollector(cluster_mode, master_address)
-                else:
+            if self._spark_sampler is not None:
+                discovered = self._spark_sampler.discover()
```

Comment:
Why do you call `discover` every time and not cache it? `discover` logs every call:
```
            self._logger.info(
                "We already know cluster mode and master address",
                cluster_mode=self._cluster_mode,
                master_address=self._master_address,
            )
```

This is bad API. Do not allow it - instead of logging, you should `assert` discover is not called after discovery succeeded, and have your caller handle the state machine; or alternatively, handle it silently yourself without any side effect such as logs etc.

Reply:
In the meantime I'll cache `discover`, it'll won't be called it returned `True` once
I'll remove the logs in a separate PR.

Reply:
https://github.com/Granulate/granulate-utils/pull/138/files#diff-e8dd154b7df986eb768fe56114e039f44a244a6a5d75f464611aa95fcb21866fL298

Reply:
caching discover and it called until returns `True`
https://github.com/Granulate/gprofiler/pull/749/files#diff-18dce65a9628cdb97bacc94da9b0401f8686d3326c5a71c91a23eb0bf4cbb919R71-R72

Reply:
Logs are not removed yet. I think a saner API is allowing `discover` to be called only once, or at the very least, ignoring futher `discover` calls after discovery succeeded once. Not required for this PR, anyway.